src/graph/graph.c

```diff
@@ -25,20 +26,31 @@ void _Graph_LeaveCriticalSection(Graph *g) {
     pthread_mutex_unlock(&g->_mutex);
 }
 
+/* Force execution of all pending operations on a matrix. */
+static inline void _Graph_SynchronizeMatrix(GrB_Matrix m) {
+    GrB_Index nvals;
+    GrB_Matrix_nvals(&nvals, m);
+}
+
 /* Resize given matrix, such that its number of row and columns
- * matches the number of nodes in the graph. */
+ * matches the number of nodes in the graph. Also, synchronize
+ * matrix to execute any pending operations. */
 void _Graph_ResizeMatrix(const Graph *g, GrB_Matrix m) {
     GrB_Index n_rows;
-
     GrB_Matrix_nrows(&n_rows, m);
-    if(n_rows != Graph_NodeCount(g)) {
+
+    // If the matrix requires has pending operations or requires
+    // a resize, enter critical section.
+    if(GxB_Matrix_Pending(m) || (n_rows != Graph_NodeCount(g))) {
         _Graph_EnterCriticalSection((Graph *)g);
-        {
-            // Double check now that we're in critical section.
-            GrB_Matrix_nrows(&n_rows, m);
-            if(n_rows != Graph_NodeCount(g))
-                assert(GxB_Matrix_resize(m, Graph_NodeCount(g), Graph_NodeCount(g)) == GrB_SUCCESS);
-        }
+        // Double-check if resize is necessary.
+        GrB_Matrix_nrows(&n_rows, m);
+        if(n_rows != Graph_NodeCount(g))
+          assert(GxB_Matrix_resize(m, Graph_NodeCount(g), Graph_NodeCount(g)) == GrB_SUCCESS);
+
+        // Flush changes to matrices if necessary.
+        if (GxB_Matrix_Pending(m)) _Graph_SynchronizeMatrix(m);
```

Comment:
Incase GxB_Matrix_resize introduces pending changes (need to verify that) it is curtail that we maintain the order of first resize and only then check for pending changes, consider adding a comment if that's the case.